doc/design.md

```diff
@@ -0,0 +1,928 @@
+- Provide two library versions, probably with a default.
+  For example, we could initially default to using the double-precision library, using the single-precision version only on demand for applications with high performance demands after the particular workflow has been validated to be correct in single precision.
+  At a later point, in case we have convinced ourselves that the single-precision version is always sufficient, the default could be changed.
+
+It is not entirely clear how much development overhead this would cause for code depending on this library.
+More code is likely to require templating for the precision (`double` and `single`).
+However, the development cost for this library itself would be rather minor, so there is very little risk involved in supporting it from the beginning.
+
+
+## <a name="implementation"></a>Implementation
+
+
+### <a name="implementation-goals-and-non-goals"></a>Goals and non-goals
+
+##### Goals in this implementation phase
+
+1. Provide a simple, lightweight, functional and well-tested library that can replace and enhance most of the workspaces types existing in Mantid.
+1. Reach a point where this library (in combination with some supporting libraries) is useful in practice.
+   We define useful as:
+   - Complete low-level functionality in C++ and Python.
+   - `numpy` interoperability.
+   - Visualization using a slice viewer.
+   - Visualization using a multi-dimensional table viewer.
+   - Instrument visualization using the Mantid instrument viewer.
+   - Can leverage existing Mantid code such as `DataHandling::LoadEventNexus` by a subsequent conversion into a `Dataset`.
+1. Demonstrate usefulness on at least two real examples, ideally bringing them to a production ready stage.
+   Candidates are:
+   - Constant-wavelength (reactor) workflows, which currently use `API::MatrixWorkspace` with length-1 histograms as a workaround.
+   - Imaging workflows, which would benefit from presenting data as a stack of images rather than a list of histograms, i.e., the dimension order would be the opposite of what we are used to in `API::MatrixWorkspace`.
+   - Workflow with parameter scan or polarization analysis to demonstrate new capabilities of extra dimensions and/or multiple data variables.
+1. If applicable, demonstrate performance gain over existing Mantid workflow.
+
+##### Non-goals in this implementation phase
+
+1. It is a non-goal to provide a drop-in replacement for all existing workspaces.
+1. It is a non-goal to convert all 970 algorithms to use `Dataset` instead of `API::Workspace`.
+1. It is a non-goal to provide support for all visualization widgets.
+
+
+### <a name="implementation-milestones"></a>Milestones
+
+*To do:
+This needs a lot of work.
+It is quite unclear what belongs to which milestone, the idea is the have something more usable at every step.
+Also, do we have any ideas that can ultimately avoid a hard and risky "drop all the old stuff" step?
+*
+
+##### Milestone 1
```

Comment:
I agree with breaking things into milestones. I also really agree with having the milestones providing some kind of useable/useful deliverable.  I guarantee that one of the first questions asked will be how many people and how long will it take. I see three options:
1. Explicitly saying you cannot estimate without knowing more
2. Provide estimates for +1 milestone ahead giving sufficient ranges of delivery times.
3. Provide estimates for everything (likely a bad idea)

Reply:
Sorry, just seen estimates below

Reply:
I think we can give relatively good estimates for what I am listing (thinkgs like the GUI/widget integration would need more research though, in particular the opinion from someone who works on those bits).

For going beyond that I agree: We cannot give estimates, and in particular I do not even know *what* should be done (and in my opinion that decision cannot be made before we have passed several of the listed milestones.